### PR TITLE
Convert2java [Draft]

### PR DESCRIPTION
**Pull request by chaws**
Converting Kotlin to Java

### CODE REVIEW

File: src/main/java/com/github/javaparser/matchersj/MatchContext.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
+package com.github.javaparser.matchersj;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Created by djc3 on 6/13/17.
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/MatchContext.java]: We should remove these comments

File: src/main/java/com/github/javaparser/matchersj/MatchResult.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import com.github.javaparser.ast.Node;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Created by djc3 on 6/13/17.
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/MatchResult.java]: Please remove

File: src/main/java/com/github/javaparser/matchersj/Matcher.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.github.javaparser.matchersj;
+
+import com.github.javaparser.ast.Node;
+
+/**
+ * Created by djc3 on 6/13/17.
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/Matcher.java]: Please remove

File: src/main/java/com/github/javaparser/matchersj/Matchers.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.matchersj.classes.*;
+import com.github.javaparser.symbolsolver.javaparser.Navigator;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Created by cdo on 6/15/17.
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/Matchers.java]: Please remove

File: src/main/java/com/github/javaparser/matchersj/classes/AllOf.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import com.github.javaparser.matchersj.MatchResult;
+import com.github.javaparser.matchersj.Matcher;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Created by djc3 on 6/14/17.
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/classes/AllOf.java]: Please remove

File: src/main/java/com/github/javaparser/matchersj/classes/AnyOf.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.matchersj.MatchContext;
+import com.github.javaparser.matchersj.MatchResult;
+import com.github.javaparser.matchersj.Matcher;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Created by cdo on 6/15/17.
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/classes/AnyOf.java]: Please remove

File: src/main/java/com/github/javaparser/matchersj/classes/HasChild.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
+package com.github.javaparser.matchersj.classes;
+
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.matchersj.MatchContext;
+import com.github.javaparser.matchersj.MatchResult;
+import com.github.javaparser.matchersj.Matcher;
+
+/**
+ * Created by cdo on 6/15/17.
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/classes/HasChild.java]: Please remove

File: src/main/java/com/github/javaparser/matchersj/classes/HasDescendant.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.matchersj.MatchContext;
+import com.github.javaparser.matchersj.MatchResult;
+import com.github.javaparser.matchersj.Matcher;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Created by cdo on 6/15/17.
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/classes/HasDescendant.java]: Please remove

File: src/main/java/com/github/javaparser/matchersj/classes/Is.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
+package com.github.javaparser.matchersj.classes;
+
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.matchersj.MatchContext;
+import com.github.javaparser.matchersj.MatchResult;
+import com.github.javaparser.matchersj.Matcher;
+
+/**
+ * Created by cdo on 6/15/17.
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/classes/Is.java]: Please remove

File: src/main/java/com/github/javaparser/matchersj/classes/Parent.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
+package com.github.javaparser.matchersj.classes;
+
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.matchersj.MatchContext;
+import com.github.javaparser.matchersj.MatchResult;
+import com.github.javaparser.matchersj.Matcher;
+
+/**
+ * Created by cdo on 6/15/17.
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/classes/Parent.java]: Please remove

File: src/test/java/com/github/javaparser/matchersj/MatchersTest.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.Assert.*;
+
+import static com.github.javaparser.matchersj.Matchers.*;
+
+/**
+ * Created by cdo on 6/15/17.
```
Comment (ftomassetti) [inline src/test/java/com/github/javaparser/matchersj/MatchersTest.java]: Please remove

File: src/test/java/com/github/javaparser/matchersj/MatchersTest.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+                                                              Parent(Is(ForeachStmt.class))
+                                                        ),
+                                                        HasDescendant(Is(MethodCallExpr.class))
+                                                    )
+                                                );
+        assertEquals(2, matches.size());
+        assertTrue(matches.stream().anyMatch(mr -> BooleanLiteralExpr.class.isInstance(mr.getCurrentNode())));
+        assertTrue(matches.stream().anyMatch(mr -> BlockStmt.class.isInstance(mr.getCurrentNode())));
+    }
+}
```
Comment (ftomassetti) [inline src/test/java/com/github/javaparser/matchersj/MatchersTest.java]: Newline missing at the end of the file

File: src/main/java/com/github/javaparser/matchersj/MatchResult.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    public N getCurrentNode() {
+        return currentNode;
+    }
+
+    public MatchResult<N> combine (MatchResult<N> other) {
+        if (this.isEmpty() || other.isEmpty()) {
+            return MatchResult.<N>empty((N)currentNode);
+        }
+
+        HashSet<MatchContext> combinations = new HashSet<MatchContext>();
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/MatchResult.java]: We could define the variable as of type `Set<MatchContext>`

File: src/main/java/com/github/javaparser/matchersj/MatchResult.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+                combinations.add(mc.combine(otherMc));
+            }
+        }
+
+        return new MatchResult(currentNode, combinations.stream()
+                                                        .filter(Objects::nonNull)
+                                                        .collect(Collectors.toList()));
+    }
+
+    public static void main(String[] args) {
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/MatchResult.java]: Please remove

File: src/main/java/com/github/javaparser/matchersj/Matchers.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+ * Most static methods in this class are a bare call to its classes
+ * This was written in order to get better code elegancy when writting matchers
+ */
+public class Matchers {
+
+    public static <N extends Node, T extends Node> Matcher<N> Is(Class<T> type) {
+        return new Is<N, T>(type);
+    }
+
+
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/Matchers.java]: Extra space

File: src/main/java/com/github/javaparser/matchersj/Matchers.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+import com.github.javaparser.symbolsolver.javaparser.Navigator;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Created by cdo on 6/15/17.
+ *
+ * Most static methods in this class are a bare call to its classes
+ * This was written in order to get better code elegancy when writting matchers
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/Matchers.java]: writting -> typo

Yes, I think these methods are a really good idea: we need those. We should just rename all of them to start with a lowercase letter

File: src/main/java/com/github/javaparser/matchersj/classes/AllOf.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+ */
+public class AllOf<N extends Node> implements Matcher<N> {
+
+    private List<Matcher<N>> elements = new ArrayList<>();
+
+    public AllOf(Matcher<N>... elements) {
+        this.elements = Arrays.asList(elements);
+        if (elements.length == 0) {
```
Comment (ftomassetti) [inline src/main/java/com/github/javaparser/matchersj/classes/AllOf.java]: This should come before the assignment

File: src/test/java/com/github/javaparser/matchersj/MatchersTest.java [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+import static org.junit.Assert.*;
+
+import static com.github.javaparser.matchersj.Matchers.*;
+
+/**
+ * Created by cdo on 6/15/17.
+ */
+public class MatchersTest {
+    @Before
+    public void setUp() throws Exception {
```
Comment (ftomassetti) [inline src/test/java/com/github/javaparser/matchersj/MatchersTest.java]: We can remove this one

File: src/test/java/com/github/javaparser/matchersj/MatchersTest.java [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    @After
+    public void tearDown() throws Exception {
```
Comment (ftomassetti) [inline src/test/java/com/github/javaparser/matchersj/MatchersTest.java]: We can remove this one

**Review by ftomassetti — COMMENTED** [review]
Thank you so much. It looks good!
I assume all the tests are passing, correct?

Once I merge this I will configure Travis to run tests automatically.

Also, it seems you have added the Java files but you did not remove the corresponding Kotlin files. I think you should remove them.

I am very happy to see this moving forward!

**Comment by chaws** [comment]
Thanks for reviewing it!
I've thumb'd up most of your comments and already pushed two commits: one removing all kotlin from the project (if you could review build.gradle just to make sure, it'd be great) and another one with containing all the other small changes.

Last test isn't working, I commented it out so I can take a look at it sometime this week, all the remaining tests are working just fine.

Cheers!

**Comment by ftomassetti** [comment]
Good, let's merge this and I will later check the test and configure Travis for Continous Integration.

Thanks!

**Comment by ftomassetti** [comment]
Ok, now the test is fixed and more have been added. Also, tests now will run automatically for each pull request

**Comment by chaws** [comment]
That's awesome, thanks for merging!

**Pull request merged**